EIPS/eip-inclusionlists.md

```diff
@@ -0,0 +1,170 @@
+
+## Specification
+
+### Constants
+
+| Name | Value |
+| - | - |
+| `MAX_TRANSACTIONS_PER_INCLUSION_LIST` | `2**4 = 16` |
+| `MAX_GAS_PER_INCLUSION_LIST` | `2**21` |
+| `MIN_SLOTS_FOR_INCLUSION_LIST_REQUEST` | `1` |
+
+#### Reference Objects
+
+```
+class InclusionListSummaryEntry(Container):
+    address: ExecutionAddress
+    gas_limit: uint64
+```
+
+```
+class InclusionListSummary(Container)
+    slot: Slot
+    proposer_index: ValidatorIndex
+    summary: List[InclusionListSummaryEntry, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
+```
+
+```
+class SignedInclusionListSummary(Container):
+    message: InclusionListSummary
+    signature: BLSSignature
+```
+
+```
+class InclusionList(Container)
+    summary: SignedInclusionListSummary
+    transactions: List[Transaction, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
+```
+
+```
+class ExecutionPayload(Container):
+    ...
+    inclusion_list_summary: List[InclusionListSummaryEntry, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
+    inclusion_list_exclusions: List[uint64, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
+```
+
+```
+class ExecutionPayloadHeader(Container):
+    ...
+    inclusion_list_summary_root: Root
+    inclusion_list_exclusions_root: Root
+```
+
+```
+class BeaconBlockBody(Container):
+    ...
+    inclusion_list_summary: SignedInclusionListSummary
+```
+
+### Consensus layer
+
+#### High-level overview
+
+**`slot N` proposal:**
+- Proposer broadcasts a signed block and an inclusion list (summary and transactions objects) for `slot N+1`.
+- Transactions will be included in `slot N` or `slot N+1`.
+- Summaries include the originating address of the transactions and their respective gas limits.
+- Summaries are signed, but transactions are not.
+
+**`slot N` validation:**
+- Validators only consider the block for validation and fork-choice if they have seen at least one inclusion list for that slot.
+- They consider the block invalid if the inclusion list transactions are not executable at the start of `slot N` or if the transactions' `maxFeePerGas` are not at least 12.5% higher than the current slot (to ensure the transactions will be valid in `slot N+1`).
```

Comment:
If sequences transactions of the same account are allowed, there's another potential issue. The `slot N` proposer could force the next proposer to include its transaction as the highly-sought first transaction in `slot N+1` for MEV purposes.  It could, for example, include the following transactions (from the same account) in a list:
1. tx1 calls a contract with all the account's eth. The contract checks the price of a volatile token.  If it equals the price at the end of `slot N`, the contract sends the eth back to msg.sender. Otherwise it doesn't.
2. tx2 does something that takes advantage of being at the beginning of the block, e.g. extract arbitrage from the above token.

If the list can contain both tx1 and tx2, then the proposer of `slot N+1` must include them before any other transaction that changes the token price.

This can be mitigated by the same rule I proposed above: restricting inclusion lists to one transaction per account, and only if the block doesn't include a transaction of that account.

Reply:
ty for the comments: 

> Would the slot N proposer be able to DoS the slot N+1 proposer (causing it to miss the slot due to not being able to produce a valid block) by including a transaction that is valid at the start of slot N but invalid at its end?

Nope! If you check out the design in https://ethresear.ch/t/no-free-lunch-a-new-inclusion-list-design/16389, we talk specifically about solving this issue! basically instead of committing to a specific transaction, the proposer commits to any transaction from that address. this helps avoid this issue. we cannot ensure that the transaction is valid in the post state of slot N because in PBS (either mev-boost or enshrined) the prooser commits to a block without knowing the contents.

> If sequences transactions of the same account are allowed, there's another potential issue.

Fortunately, this isn't an issue either! Because the IL just commits to a from address, there can only be one IL transaction per-address per-slot!

Reply:
Thanks for clarifying.  Listing the sender rather than the transaction solves all of these issues.  If block N contains any transaction from that address, the condition is satisfied so it can't invalidate another transaction that obligates N+1.  Neat!

IL becomes a bigger challenge once we introduce AA.  Accounts can invalidate each other's transactions.  There are mempool rules to protect the builder from DoS vectors introduced by this trait, but they won't solve the issue of proposer N attacking N+1 by such a cross invalidation.  I've been chatting with Barnabé about this.  It's something we'll need to figure out.

Reply:
for sure! thanks for bringing this perspective :-) def need to ensure that the IL mechanism we enshrine fits with 4337 and AA generally!